pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/ClassNamingConventionsRule.java

```diff
@@ -4,15 +4,144 @@
+import net.sourceforge.pmd.properties.RegexProperty.RegexPBuilder;
+
 
+/**
+ * Configurable naming conventions for type declarations.
+ */
 public class ClassNamingConventionsRule extends AbstractJavaRule {
 
+    private static final RegexProperty CLASS_REGEX = defaultProp("class").desc("Regex which applies to concrete class names").build();
+    private static final RegexProperty ABSTRACT_CLASS_REGEX = defaultProp("abstract class").build();
+    private static final RegexProperty INTERFACE_REGEX = defaultProp("interface").build();
+    private static final RegexProperty ENUMERATION_REGEX = defaultProp("enum").build();
+    private static final RegexProperty ANNOTATION_REGEX = defaultProp("annotation").build();
+    private static final RegexProperty UTILITY_CLASS_REGEX = defaultProp("utility class").defaultValue("[A-Z][a-zA-Z]+Util").build();
+
+
+    public ClassNamingConventionsRule() {
+        definePropertyDescriptor(CLASS_REGEX);
+        definePropertyDescriptor(ABSTRACT_CLASS_REGEX);
+        definePropertyDescriptor(INTERFACE_REGEX);
+        definePropertyDescriptor(ENUMERATION_REGEX);
+        definePropertyDescriptor(ANNOTATION_REGEX);
+        definePropertyDescriptor(UTILITY_CLASS_REGEX);
+
+        addRuleChainVisit(ASTClassOrInterfaceDeclaration.class);
+        addRuleChainVisit(ASTEnumDeclaration.class);
+        addRuleChainVisit(ASTAnnotationTypeDeclaration.class);
+    }
+
+
+    private void checkMatches(ASTAnyTypeDeclaration node, PropertyDescriptor<Pattern> regex, Object data) {
+        if (!getProperty(regex).matcher(node.getImage()).matches()) {
+            addViolation(data, node, new Object[]{
+                    isUtilityClass(node) ? "utility class" : node.getTypeKind().getPrintableName(),
+                    node.getImage(),
+                    getProperty(regex).toString(),
+            });
+        }
+    }
+
+
+    // This could probably be moved to ClassOrInterfaceDeclaration
+    // to share the implementation and be used from XPath
+    private boolean isUtilityClass(ASTAnyTypeDeclaration node) {
+        if (node.getTypeKind() != TypeKind.CLASS) {
+            return false;
+        }
+
+        // A class without declarations shouldn't be reported
+        boolean hasAny = false;
+
+        // we could probably enrich ASTAnyTypeBodyDeclaration
+        // with methods to know what kind of declaration it is
+        for (ASTAnyTypeBodyDeclaration decl : node.getDeclarations()) {
+            AccessNode accessNode = decl.getFirstChildOfType(AccessNode.class);
+            ASTInitializer initializer = decl.getFirstChildOfType(ASTInitializer.class);
+            if (accessNode != null && !(accessNode instanceof ASTConstructorDeclaration)) {
+                hasAny = true;
+                if (!accessNode.isStatic()) {
+                    return false;
+                }
+            }
+
+            if (initializer != null && !initializer.isStatic()) {
+                return false;
```

Comment:
if the class has a constructor we ignore it, but non-static initializer aren't? I'm not sure with the criteria...

Reply:
Yes it's probably more consistent to ignore both. I assumed that non-static initialisers should be treated specially because they're used so rarely, but that's probably not a good reason

Reply:
ouch! I was reading this wrong, it was actually ok. I'll revert as I merge